python_modules/libraries/dagster-aws/dagster_aws_tests/pipes_tests/test_pipes.py

```diff
@@ -504,6 +504,34 @@ def test_cloudwatch_logs_reader(cloudwatch_client: "CloudWatchLogsClient", capsy
     assert capsys.readouterr().out.strip() == "1\n2\n3"
 
 
+def test_cloudwatch_logs_reader_does_not_repeat_logs(cloudwatch_client: "CloudWatchLogsClient", capsys):
+    reader = PipesCloudWatchLogReader(client=cloudwatch_client)
+
+    is_session_closed = threading.Event()
+
+    log_group = "/pipes/tests"
+    log_stream = "test-cloudwatch-logs-reader"
+
+    cloudwatch_client.create_log_group(logGroupName=log_group)
+    cloudwatch_client.create_log_stream(logGroupName=log_group, logStreamName=log_stream)
+
+    for i in range(5):
+        cloudwatch_client.put_log_events(
+            logGroupName=log_group,
+            logStreamName=log_stream,
+            logEvents=[
+                {"timestamp": int(datetime.now().timestamp() * 1000), "message": f"{i}"},
+            ],
+        )
+        time.sleep(0.1)
+
+    reader.start({"log_group": log_group, "log_stream": log_stream}, is_session_closed)
+    time.sleep(0.1)
+    reader.stop()
+
+    assert capsys.readouterr().out.strip() == "0\n1\n2\n3\n4"
```

Comment:
what I think _may_ be needed to make this test actually fail is the ability to log empty events, unfortunately `logEvents` is required on `put_log_events()` the reason I say that is because when I print out the responses from `get_log_events()` there are many with empty events, e.g.

![Screenshot 2024-12-10 at 16 44 10](https://github.com/user-attachments/assets/2500993d-0676-491e-9e01-9e57aebf9676)

A theory I have is that `tail_cloudwatch_events()` is the issue due to events often being an empty, although I've not been able to prove that thus far.

Reply:
Interesting, but I don't see how empty events would cause duplicated logs. 
Also, I don't think empty events are even being sent to stdout/stderr.